# ANNOVAR #

Thank you for your interests in using ANNOVAR. An updated version (2014nov12) is
available.  

Here is the link to download ANNOVAR:  
http://www.openbioinformatics.org/annovar/download/g4EUwyphi9/annovar.latest.tar.gz  

Please feel free to contact me if you have any questions or comments, but do NOT
reply this email as it will be IGNORED! Use a different title in your email.  

Please cite ANNOVAR if you use it in your research (Wang K, Li M, Hakonarson H.
ANNOVAR: Functional annotation of genetic variants from next-generation
sequencing data, Nucleic Acids Research, 38:e164, 2010). I spent tremendous
amount of time and efforts to maintain this tool, and your citation really means
a lot to me.  

### ADDITIONAL NOTES TO ANNOVAR USERS: ###  

Several notable improvements in the new version of ANNOVAR include: (1)
significantly reduce memory usage for filter annotation (2) improve
compatibility for unconventional chromosome names for species such as tomato (3)
fixed a problem in exon numbering for splice variants in reverse strand (4)
other minor functional improvements and database updates.  

Several notable additions/updates to ANNOVAR databases include: (1) updated
refGene, knownGene and ensGene files for hg18/hg19/hg38 are available to
download (2) 1000 Genomes Project 2014Oct version is available for all
populations and five subpopulations (AFR,AMR,EAS,EUR,SAS) with chrX/Y markers.
(3) ExAC 65000 exomes v0.2 is available all all populations and seven
subpopulations (AFR,AMR,EAS,FIN,NFE,OTH,SAS). (4) Updated ljb26 databases from
dbNSFP indexed by ANNOVAR is available to download. (5) Updated Clinvar and
Cosmic are available to download.  

Point of interest to ANNOVAR users: (1) A recent Genome Biology paper
(http://genomebiology.com/2014/15/3/R53) demonstrated that ANNOVAR is the most
widely used tool (>50% share) for interpretation of clinical genome sequencing
data (2) A recent Genome Medicine paper
(http://genomemedicine.com/content/6/3/26) reported that the choice of
transcript set can have a large effect on the ultimate variant annotations from
ANNOVAR and VEP (ENSEMBLâ€™s Variant Effect Predictor). (3) Multiple users asked
my comments to a blog (http://blog.goldenhelix.com/?p=2486). I recently posted a
reply at the end of the page, clarifying that ANNOVAR does generate correct and
biologically meaningful results in all examples shown in the blog. (4) Some
users may not be aware of the wANNOVAR server (http://wannovar.usc.edu) which
allows annotation of VCF files via a web interface. Compared to ANNOVAR, it has
limited functionality and the data sources may not be as updated; however, it is
more user friendly.  

Mailing list related issues: (1) I typically send out an update email
every 4 months. (2) if you wish to be removed from the list, please email me
directly with "unsubscribe ANNOVAR" in the title of the email. (3) Starting from
next update, I may send emails to user via professional mailing list, rather
than from my personal email account.  

Finally, thank you very much for your help and support to ANNOVAR!  